### ISSUE TITLE
How to build voicevox_core manually to work with model, which is inside prebuilt release version?

### ISSUE THREAD

**Issue by jaehwa-v6x** [VOICEVOX/voicevox_core#492]
## 質問の内容

First of all, thank you for creating such an awesome project!

I would like to use voicevox_core in aws lambda.
However, the existing prebuilt version voicevox_core-linux-x64-cpu-0.14.3.zip is built on Ubuntu 20, which requires glibc version 2.29 or higher.
On the other hand, the runtime for aws lambda(amazonlinux2), is currently shipped with glibc 2.26, so we can't load prebuilt voicevox core library directly from aws lambda.

Maybe we can use prebuilt version by loading aws lambda as a Docker image, but I want to avoid this for various reasons, so I would like to build by myself and use it with model inside prebuilt release version.

However, there is a dummy model in src/status/model.rs, so when I compile libvoicevox_core.so file and load it with the model that is inside prebuilt version, it didn't load.

Perhaps I need to modify src/status/model.rs to match the model of prebuilt version and then build it, is my understanding correct? (I didn't find it with previous issues).

MODEL_FILE_NAMES doesn't seem to be too difficult, but I still haven't figured out how to configure SPEAKER_ID_MAP, so is it possible to get some guideline or hints to do that? Or if I can get model.rs which is used to build prebuilt package, it would be great!

## VOICEVOXのバージョン

0.13.0

## OSの種類/ディストリ/バージョン

- [ ] Windows
- [ ] macOS
- [X] Linux

Linux ubuntu 18 64bit

## その他

I'm impressed with the flexible and tightly knit package, thanks again!

**Comment by sevenc-nanashi** [comment]
TL;DR: You can't, characters' model is proprietary.

Verbosely:
Many characters restricts about using their character for political use, sexual use (I think), and etc. This is not compatible with OSS License. So characters' model is proprietary, and models which is included in repository are dummy.

----

I heard that Ubuntu 18 will be EOL soon, so I think that it's the best way to upgrade your Ubuntu. (idk it's possible in AWS lambda)

<details>
<summary>日本語</summary>
ざっくり：無理です。製品版のモデルはプロプライエタリです。

詳細：
多くのキャラクターは政治的、性的（多分）な使用を禁じています。これはOSSライセンスと合いません。
そのため、キャラクターのモデルはプロプライエタリにして、リポジトリのモデルはダミーになっています。

Ubuntu 18のEOLが近いらしいので、Ubuntuを更新するのが一番だと思います（AWS Lambdaで可能かどうかはわかりませんが）

</details>

**Comment by Hiroshiba** [comment]
@jaehwa-v6x Thank you for creating the issue!

As @sevenc-nanashi mentioned, we currently do not support builds that include character information.

I've done a bit of research and it seems like AWS Lambda utilizes Amazon Linux 2, and updating glibc appears to be a difficult task... There appears to be a new Amazon Linux 2022 in development, but it doesn't seem available for use from Lambda yet.
Given this situation, it seems unlikely to run VOICEVOX Core smoothly on AWS Lambda. On the other hand, it would also be quite a task for us at VOICEVOX to support an older version of glibc... If only there were a good solution... ;-)

<details>
<summary>日本語</summary>

@sevenc-nanashi のコメントの通り、キャラクター情報を含んだビルドは今のところサポートされていません。

ちょっと調べたのですが、AWS LambdaはAmazon Linux 2が使われていて、かつglibcのアップデートなどは難しそうなんですね･･･。
新しくAmazon Linux 2022というのが作られていそうですが、まだLambdaから利用できないように見えました。

となると、VOICEVOX Coreは簡単に動かせないということになりそうです。
かといってVOICEVOX側で古いglibcをサポートするのも大変ですし･･･。なにか良い解決策があれば良いのですが･･･。 ;-)

</details>

**Comment by Patchethium** [comment]
@sevenc-nanashi is right, use docker instead. The problem is that the neural models with actual working weight is never released, modifying Rust code won't help.

You might be able to switch your OS version, or get an extra copy by contacting the owner, but they're not worth the effort. Just use docker, that's what it's born for.

BTW I've always been confused how many English speakers are around here for a Japanese TTS, maybe you could tell me the reason you choose this repo instead of Elevenlabs or coqui studio?

**Comment by jaehwa-v6x** [comment]
Thanks for the quick response!

Can you confirm if I understand you correctly? It is possible to use pre-built binaries and models by putting them on my server (Under the restriction of each character and voice). However, building my own binary including models inside pre-built release is not supported due to concerns of misuse. Right?

@Patchethium  I am looking for a decent Japanese TTS voice. The quality of VoiceVox is very high and popular. :)

**Comment by Patchethium** [comment]
Thanks for the explanation, yes you've got it right, you can run it, but can't modify or reproduce one.

**Comment by jaehwa-v6x** [comment]
> > Maybe we can use prebuilt version by loading aws lambda as a Docker image, but I want to avoid this for various reasons
> 
> If you don't mind, could you tell us the reasons ?

During development, I mostly work on a Linux desktop, but I'd like to run it locally on my MacBook as well, and Docker only runs on Linux by default. Another reason is that I have functions that I've already written, and mixing them with Docker seems complicated, so I'll probably have to move that code to a Docker image as well. However, a Docker image is a viable option, so I'll give it a try.

> > However, building my own binary including models inside pre-built release is not supported due to concerns of misuse. Right?
> 
> Yes. "VOICEVOX" needs to be proprietary, and VOICEVOX CORE is the proprietary part.
> 
> edit: I think you saw `fn decrypt` in model_file.rs. It is the proprietary part of VOICEVOX CORE.

I completely understand now, thanks for the clarification!

**Comment by finaltrain** [comment]
お世話になっております。Voicevoxを使わせていただいているプログラミング初心者です。

現在、「コアライブラリのビルド」の項目に「model フォルダにある onnx モデルはダミーのため、ノイズの混じった音声が出力されます」とありますが、こちらのissueにたどり着くまでその意味が「セルフコンパイルした場合はダミーモデルしか使えず、キャラクターボイスを使うことができない」ということであるのを理解することができませんでした。
「全体構成について」も拝見しましたが、そこでも理解できませんでした。

jaehwa-v6xさんや私のようにプレビルド版のmodelデータを読み込めるようにセルフコンパイルしようとしているユーザーは他にもいるかと思われます。どこかにこのことを明確に記載していただけると、初心者には分かりやすく助けになるかと思います。

ドキュメントを整理したい系のissueが複数ありどこに書くのが適切か迷ったため、こちらに記載しておきます。
お忙しいところ恐れ入りますが、よろしくお願いいたします。

現状glibcが古いlinuxで動かしたかったらdocker使うのがよさそうですね。

**Comment by Hiroshiba** [comment]
@finaltrain コメントありがとうございます！

今リードミーを確認したのですが、おっしゃる通りかなり分かりづらくなっているように感じました。
特にここの文章は「Releasesにあるコアライブラリは自分で１からビルドすることができる」と受け取れてしまいそうです。

> Releases にあるビルド済みのコアライブラリを利用せず、自分で一からビルドする場合こちらを参照してください。ビルドには Rust (Windows での Rust 開発環境構築手順はこちら) と cmake が必要です。
> 
> model フォルダにある onnx モデルはダミーのため、ノイズの混じった音声が出力されます

こちらをこのように書き換えると分かりやすそうでしょうか？

---

ビルドには [Rust](https://www.rust-lang.org/ja) ([Windows での Rust 開発環境構築手順はこちら](https://docs.microsoft.com/ja-jp/windows/dev-environment/rust/setup)) と [cmake](https://cmake.org/download/) が必要です。
[Releases](https://github.com/VOICEVOX/voicevox_core/releases) にあるビルド済みのコアライブラリを利用せず、自分で一からビルドした場合は、model フォルダにある onnx モデルのみが読み込めます。
このモデルはダミーのため、ノイズの混じった音声が出力されます。

**Comment by finaltrain** [comment]
@Hiroshiba
ご対応ありがとうございます！
これなら分かりやすいです！
